src/index.cpp

```diff
@@ -1652,57 +1651,77 @@ namespace diskann {
       return -2;
     }
 
-    std::unique_lock<std::shared_timed_mutex> update_lock(_update_lock);
+    std::unique_lock<std::shared_timed_mutex> ul(_update_lock);
+    std::unique_lock<std::shared_timed_mutex> tl(_tag_lock);
+    std::unique_lock<std::shared_timed_mutex> dl(_delete_lock);
+
     if (_data_compacted) {
       for (unsigned slot = (unsigned) _nd; slot < _max_points; ++slot) {
         _empty_slots.insert(slot);
       }
     }
 
-    _lazy_done = false;
     return 0;
   }
 
   template<typename T, typename TagT>
   inline void Index<T, TagT>::process_delete(
-      const tsl::robin_set<unsigned> &old_delete_set, size_t i,
-      const unsigned &range, const unsigned &maxc, const float &alpha,
+      const tsl::robin_set<unsigned> &old_delete_set, size_t loc,
+      const unsigned range, const unsigned maxc, const float alpha,
       InMemQueryScratch<T> *scratch) {
-    tsl::robin_set<unsigned> candidate_set;
-    std::vector<Neighbor>    expanded_nghrs;
+    tsl::robin_set<unsigned> expanded_nodes_set;
+    std::vector<Neighbor>    expanded_nghrs_vec;
 
-    bool modify = false;
+    // If this condition were not true, deadlock could result
+    assert(old_delete_set.find(loc) == old_delete_set.end());
 
-    for (auto ngh : _final_graph[(_u32) i]) {
-      if (old_delete_set.find(ngh) != old_delete_set.end()) {
+    std::vector<unsigned> adj_list;
+    {
+      // Acquire and release lock[loc] before acquiring locks for neighbors
+      std::unique_lock<non_recursive_mutex> adj_list_lock;
+      if (_conc_consolidate)
+        adj_list_lock = std::unique_lock<non_recursive_mutex>(_locks[loc]);
+      adj_list = _final_graph[loc];
+    }
+
+    bool modify = false;
+    for (auto ngh : adj_list) {
+      if (old_delete_set.find(ngh) == old_delete_set.end()) {
+        expanded_nodes_set.insert(ngh);
+      } else {
         modify = true;
 
-        // Add outgoing links from
+        std::unique_lock<non_recursive_mutex> ngh_lock;
         if (_conc_consolidate)
-          _locks[ngh].lock();
+          ngh_lock = std::unique_lock<non_recursive_mutex>(_locks[ngh]);
         for (auto j : _final_graph[ngh])
-          if (old_delete_set.find(j) == old_delete_set.end())
-            candidate_set.insert(j);
-        if (_conc_consolidate)
-          _locks[ngh].unlock();
-      } else {
-        candidate_set.insert(ngh);
+          if (j != loc && old_delete_set.find(j) == old_delete_set.end())
+            expanded_nodes_set.insert(j);
       }
     }
     if (modify) {
-      for (auto j : candidate_set) {
-        expanded_nghrs.push_back(
-            Neighbor(j, _distance->compare(_data + _aligned_dim * i,
-                                           _data + _aligned_dim * (size_t) j,
-                                           (unsigned) _aligned_dim)));
+      std::unique_lock<non_recursive_mutex> adj_list_lock(_locks[loc]);
```

Comment:
We're holding this lock longer than needed for the `else` branch. Compared to previous commits in the PR, we're keeping the lock while computing the distances and sorting `expanded_nghrs_vec`, which is unnecessary.

I'd suggest moving the lock and `_final_graph[loc].clear();` call into both `if` and `else` branch (in the `else branch, the lock and the `clear` call would go just before the `occlude_list` call).

Reply:
In the else code path, wouldn't most time be spend on occlude_list? If you dont want that overhead, we should create a new buffer pass it to occlude_list and then acquire lock to copy to the _final_graph

Reply:
Synced offline, that can be another step further.